xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+    final XdsClientWrapperForServerSds xdsClientWrapperForServerSds =
+        createXdsClientWrapperForServerSds(port, downstreamTlsContext);
+    SdsProtocolNegotiators.ServerSdsProtocolNegotiator serverSdsProtocolNegotiator =
+        new SdsProtocolNegotiators.ServerSdsProtocolNegotiator(xdsClientWrapperForServerSds);
+    server = cleanupRule.register(builder.buildServer(serverSdsProtocolNegotiator)).start();
+    return xdsClientWrapperForServerSds.getListenerWatcher();
+  }
+
+  /** Creates XdsClientWrapperForServerSds: also used by other classes. */
+  public static XdsClientWrapperForServerSds createXdsClientWrapperForServerSds(
+      int port, DownstreamTlsContext downstreamTlsContext) {
+    XdsClient mockXdsClient = mock(XdsClient.class);
+    XdsClientWrapperForServerSds xdsClientWrapperForServerSds =
+        new XdsClientWrapperForServerSds(port, mockXdsClient, null);
+    createListenerUpdate(port, downstreamTlsContext,
+        xdsClientWrapperForServerSds.getListenerWatcher());
+    return xdsClientWrapperForServerSds;
+  }
+
+  private static int findFreePort() throws IOException {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      socket.setReuseAddress(true);
+      return socket.getLocalPort();
+    }
+  }
+
+  private static void createListenerUpdate(
+      int port, DownstreamTlsContext tlsContext, XdsClient.ListenerWatcher registeredWatcher) {
+    EnvoyServerProtoData.Listener listener =
+        buildListener("listener1", "0.0.0.0", port, tlsContext);
+    XdsClient.ListenerUpdate listenerUpdate =
+        XdsClient.ListenerUpdate.newBuilder().setListener(listener).build();
+    registeredWatcher.onListenerChanged(listenerUpdate);
+  }
+
+  static EnvoyServerProtoData.Listener buildListener(
+      String name,
+      String address,
+      int port,
+      DownstreamTlsContext tlsContext) {
+    EnvoyServerProtoData.FilterChainMatch filterChainMatch =
+        buildFilterChainMatch(port, address);
+    EnvoyServerProtoData.FilterChain filterChain1 =
+        new EnvoyServerProtoData.FilterChain(filterChainMatch, tlsContext);
+    EnvoyServerProtoData.Listener listener =
+        new EnvoyServerProtoData.Listener(name, address, Arrays.asList(filterChain1));
+    return listener;
+  }
+
+  private void buildClientAndTest(
+      UpstreamTlsContext upstreamTlsContext,
+      String overrideAuthority,
+      String requestMessage,
+      int serverPort) {
+
+    XdsChannelBuilder builder =
+        XdsChannelBuilder.forTarget("localhost:" + serverPort).tlsContext(upstreamTlsContext);
+    if (overrideAuthority != null) {
+      builder = builder.overrideAuthority(overrideAuthority);
+    }
+    SimpleServiceGrpc.SimpleServiceBlockingStub blockingStub =
+        SimpleServiceGrpc.newBlockingStub(cleanupRule.register(builder.build()));
+    String resp = unaryRpc(requestMessage, blockingStub);
+    assertThat(resp).isEqualTo("Hello " + requestMessage);
```

Comment:
this doesn't verify any security, it just verifies it is broken or not. is there way to verify?

Reply:
The intent is not to verify the actual TLS handshake but that an RPC was successful after an xDS/SDS managed mTLS channel was set up: so it is more of an E2E test without actually setting up xDS or even SDS server. So this part of the test hasn't changed

Reply:
BTW, security is verified by using a "bad" cert which is part of `mtlsClientServer_changeServerContext_expectException`. That also tests updating the server context.